lib/ansible/modules/web_infrastructure/rundeck_project.py

```diff
@@ -0,0 +1,193 @@
+        required: True
+    url:
+        description:
+            - Sets the rundeck instance URL.
+        required: True
+    api_version:
+        description:
+            - Sets the API version used by module.
+            - API version must be at least 14.
+        default: 14
+    token:
+        description:
+            - Sets the token to authenticate against Rundeck API.
+        required: True
+
+
+'''
+EXAMPLES = '''
+- name: Create a rundeck project
+  rundeck_project:
+    name: "Project_01"
+    api_version: 18
+    url: "https://rundeck.example.org"
+    token: "mytoken"
+    state: present
+
+- name: Remove a rundeck project
+  rundeck_project:
+    name: "Project_02"
+    url: "https://rundeck.example.org"
+    token: "mytoken"
+    state: absent
+'''
+
+RETURN = '''
+rundeck_response:
+    description: Rundeck response when a failure occurs
+    returned: failed
+    type: string
+before:
+    description: dictionnary containing project informations before modification
+    returned: success
+    type: dictionnary
+after:
+    description: dictionnary containing project informations after modification
+    returned: success
+    type: dictionnary
+'''
+
+# import module snippets
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import fetch_url
+import json
+
+
+class RundeckProjectManager:
+    def __init__(self, module):
+        self.module = module
+
+    def handle_http_code_if_needed(self, infos):
+        if infos["status"] == 403:
+            self.module.fail_json(msg="Token not allowed. Please ensure token is allowed or has the correct "
+                                      "permissions.", rundeck_response=infos["body"])
+        elif infos["status"] >= 500:
+            self.module.fail_json(msg="Fatal Rundeck API error.", rundeck_response=infos["body"])
+
+    def request_rundeck_api(self, query, data=None, method="GET"):
+        resp, info = fetch_url(self.module,
+                               "%s/api/%d/%s" % (self.module.params["url"], self.module.params["api_version"], query),
+                               data=json.dumps(data),
+                               method=method,
+                               headers={
+                                   "Content-Type": "application/json",
+                                   "Accept": "application/json",
+                                   "X-Rundeck-Auth-Token": self.module.params["token"]
+                               })
+
+        self.handle_http_code_if_needed(info)
+        if resp is not None:
+            resp = resp.read()
+            if resp != "":
+                try:
+                    json_resp = json.loads(resp)
+                    return json_resp, info
+                except ValueError as e:
+                    self.module.fail_json(msg="Rundeck response was not a valid JSON. Exception was: %s. "
+                                              "Object was: %s" % (str(e), resp))
```

Comment:
Note that str(e) can traceback if the string representation of e contains unicode characters.  (This often happens when user data is used in the exception message).  It's better to use one of our text functions to handle the conversion:
``` python
from ansible.module_utils._text import to_native
[...]
"Object was: %s" % (to_native(e), resp)
```